### ISSUE TITLE
Sessions on the API

### ISSUE THREAD

**Issue by Primer42** [tumblr/collins#317]
@mattikus started a discussion in the IRC channel about making the API support sessions.  We already have sessions on the website, with cookies, and making the API support them doesn't have any immediate problems that I can think of.
What do other people think? @byxorna @maddalab @defect @roymarantz

**Comment by mattikus** [comment]
I agree that the access should be simple and should be easy to use with cURL from the command line. This shouldn't invalidate the current basic auth and should work alongside it, allowing you to opt in as necessary. cURL should also support saving session cookies and reusing them, so you could use either basic auth or sessions on the command line trivially.

I mainly want this so we can use bulk imports or uploads from a script without needing to pass a new auth every time, since our LDAP backend uses an OTP solution. I'd like to have actual audit logs rather than rely on service accounts, if possible.

**Comment by roymarantz** [comment]
I think API sessions are fine as long as they are implemented in a way to not weaken the RESTfullness or security of the API.  I would agree (and even argue) that they help with security, as in this case where they make OTP easier/practical to use.

**Comment by defect** [comment]
How about using a token based authentication scheme? A user who wishes to do some automated task logs in and generates a token which can then be used in the requests. This to me seems a bit simpler than using sessions/cookies and having to deal with a cookie store and things like that. It would also be simple to revoke tokens.

I'm not too familiar with the auth code in collins/play, but it seems it might fit into `SecureApiController` as a second match case?

**Comment by mattikus** [comment]
I suggested sessions based on perceived ease of implementation, since they're already there for the web view stuff. A token would also work well, especially if it can be passed as basic auth. Anything that would allow me to verify who I am first and then reuse something for a limited amount of time after that.

**Comment by maddalab** [comment]
Do you need to use OTP in automation? Can you use mixed mode authentication where API clients use file based auth.?

Finally [here] (http://restcookbook.com/Basics/loggingin/) are a few options, why not use oauth2?

**Comment by mattikus** [comment]
I'd like to have audit history of who made a change. I could create service accounts which do not use OTP but then I lose the history of who made a change. Mixed mode authentication might end up being where I head just out of simplicity sake.

oauth is complicated and adds yet another dependency or code for people attempting to implement an API client. The API is attempting to be simple and straightforward. You also mention on your documentation that you will never implement it :P 

However, I'm fine with any scheme that allows me to authenticate once and then reuse some form of token or session to continue to make changes for a limited period of time from that point on.

**Comment by roymarantz** [comment]
To me the main decision is if the API should support sessions.  If there is some kind of plugable support, people can implement what ever kind they like.

**Comment by maddalab** [comment]
@mattikus How about the following?

* API end points to add and remove OAuth2 tokens
    * Requires a schema change.
    * Updates to Collins client gem
* Updates to the authentication support on API to use either basic auth or OAuth

For your use case, your code can

* Obtain a OTP
* Use OTP to generate an OAuth2 token
* Perform operations with token
* Remove OAuth2 token

**Comment by mattikus** [comment]
I'm not wild about oauth as it requires extra dependencies on my projects
which would interact with the API but if you as the maintainers feel its
the best way to achieve a token authentication mechanism I can deal with
that.

That workflow sounds okay, as well. Though I'd also suggest adding some
kind of automatic revocation of the token based on some time length as well.

**Comment by maddalab** [comment]
@mattikus I have no strong preference. If you want to try a couple of options by implementing them to see what works best, go for it, if not I'll do it when I have some time.

**Comment by mattikus** [comment]
Sounds good. I appreciate you looking into this.

**Comment by mattikus** [comment]
So I did just a *very* basic test of pulling the session auth block from the web controller `authenticate` method into the api controller and it seems like it works just fine. There's clearly some violation of DRY here, but at least it looks feasible as a starting point.

```patch
diff --git a/app/collins/controllers/Security.scala b/app/collins/controllers/Security.scala
index 54c0a2a..a2c22a4 100644
--- a/app/collins/controllers/Security.scala
+++ b/app/collins/controllers/Security.scala
@@ -131,8 +131,19 @@ trait SecureApiController extends SecureController {
   override def authenticate(request: RequestHeader): Option[User] = {
     request.headers.get(HeaderNames.AUTHORIZATION) match {
       case None =>
-        logger.debug("Got API request with no auth header")
-        None
+        logger.debug("Got API request with no auth header, trying session")
+        User.fromMap(request.session.data) match {
+          case Some(user) => user.isAuthenticated match {
+            case true =>
+              Some(user)
+            case false =>
+              logger.debug("SecureApiController.authenticate: user found, not authenticated")
+              None
+          }
+          case None =>
+            logger.debug("SecureApiController.authenticate: user not found, session data not found")
+            None
+      }
       case Some(header) =>
         try {
           parseAuthHeader(header) match {
```

**Comment by maddalab** [comment]
sweet, how does it look on the client? How did you make the call /  store the session data on the client?

**Comment by mattikus** [comment]
My client at the moment is just cURL, so the workflow goes:

    curl -c /tmp/collins_session http://localhost:8080/login \
        -d username=blake -d password=admin:first

Then you simply pass that file to future connections:

    curl -b /tmp/collins_session http://localhost:8080/api/assets

**Comment by maddalab** [comment]
Any updates matt, do you need any help? I'm assuming you are still planning to do this, right?

**Comment by mattikus** [comment]
Hey Bhaskar,

I got a bit busy with other work projects so my time to look into Collins is fairly low right now. I'm starting to think the project won't fit my needs, based on some design decisions and I don't really want to rock y'alls boat too much.

We can close this issue. I've got a direction going forward if I want to revisit this auth scenario and I can send a PR if I find that my changes are useful.
